src/main/java/org/ltr4l/nn/Ranker.java

```diff
@@ -34,4 +38,29 @@ public void writeModel(Properties prop){
   }
   public abstract double predict(List<Double> features);
 
+  protected abstract void readModel(String model);
+
+  private static String makeRegex(String regex, int num){
+    String splitter = "";
+    for (int i = 0; i < num; i++){
+      splitter = splitter + regex;
+    }
+    return splitter;
+  }
+
+  public static List<Object> toList(String line, int dim){
+    List<Object> model = new ArrayList<>();
+    if (dim > 1){
+      String splitter = makeRegex("]", dim - 1) + ", " + makeRegex("[", dim - 1);
+      String[] elements = line.split(Pattern.quote(splitter));
+      for (String elem : elements){
+        model.add(new ArrayList<>(toList(elem, dim - 1)));
+      }
+    }
+    if (dim == 1){
```

Comment:
if(dim==1){} block that just follows if(dim>1){} block looks strange. How about rephrasing it like this:

```java
assert(dim >= 1);
if(dim == 1){ ... }
else if(dim > 1){ ... } // it's okay for me to use else{ ... } instead of else if here
```